src/OpenApi/src/Transformers/Implementations/OpenApiSchemaReferenceTransformer.cs

```diff
@@ -0,0 +1,137 @@
+        foreach (var pathItem in document.Paths.Values)
+        {
+            for (var i = 0; i < OpenApiConstants.OperationTypes.Length; i++)
+            {
+                var operationType = OpenApiConstants.OperationTypes[i];
+                if (pathItem.Operations.TryGetValue(operationType, out var operation))
+                {
+                    if (operation.Parameters is not null)
+                    {
+                        foreach (var parameter in operation.Parameters)
+                        {
+                            parameter.Schema = ResolveReferenceForSchema(parameter.Schema, schemasByReference);
+                        }
+                    }
+
+                    if (operation.RequestBody is not null)
+                    {
+                        foreach (var content in operation.RequestBody.Content)
+                        {
+                            content.Value.Schema = ResolveReferenceForSchema(content.Value.Schema, schemasByReference);
+                        }
+                    }
+
+                    if (operation.Responses is not null)
+                    {
+                        foreach (var response in operation.Responses.Values)
+                        {
+                            if (response.Content is not null)
+                            {
+                                foreach (var content in response.Content)
+                                {
+                                    content.Value.Schema = ResolveReferenceForSchema(content.Value.Schema, schemasByReference);
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        document.Components ??= new OpenApiComponents();
+        document.Components.Schemas ??= new Dictionary<string, OpenApiSchema>();
+
+        foreach (var (schema, referenceId) in schemasByReference.OrderBy(kvp => kvp.Value))
+        {
+            if (referenceId is not null)
+            {
+                document.Components.Schemas[referenceId] = ResolveReferenceForSchema(schema, schemasByReference, skipResolution: true);
+            }
+        }
+
+        return Task.CompletedTask;
+    }
+
+    internal static OpenApiSchema? ResolveReferenceForSchema(OpenApiSchema? schema, ConcurrentDictionary<OpenApiSchema, string?> schemasByReference, bool skipResolution = false)
+    {
+        if (schema is null)
+        {
+            return schema;
+        }
+
+        if (!skipResolution && schemasByReference.TryGetValue(schema, out var referenceId) && referenceId is not null)
+        {
+            return new OpenApiSchema { Reference = new OpenApiReference { Type = ReferenceType.Schema, Id = referenceId } };
+        }
+
+        if (schema.AllOf != null)
+        {
+            for (var i = 0; i < schema.AllOf.Count; i++)
+            {
+                schema.AllOf[i] = ResolveReferenceForSchema(schema.AllOf[i], schemasByReference);
```

Comment:
Naively, I would have thought we needed to propagate `skipResolution`.

Reply:
The `skipResolution` parameter is only set to true when we resolve inline schemas to references in the top-level components cache to avoi recursive references. For example, if I have:

```json
components: {
	schemas: {
		Todo: {
			properties: {
				id: { type: "integer" }
			}
		}
	}
}
```

I want:

```json
components: {
	schemas: {
		Todo: {
			properties: {
				id: { $ref: "Integer" }
			}
		}
	}
}
```

not:

```json
components: {
	schemas: {
		Todo: { $ref: "Todo" }
	}
}
```

Reply:
To confirm: we never use references for top-level schemas and that's why `skipResolution` is only (and always?) true for the top-level recursive call?

Reply:
Correct.

Reply:
Given that's the case, I prefer to name `skipResolution` to `isTopLevel` and then add a comment near `if (!isTopLevel && schemasByReference ...` explaining why we should skip resolution on the top level and nowhere else.